src/views/messages-view.jsx

```diff
@@ -793,6 +836,125 @@ class MessagesView extends React.Component {
     uploader.cancel();
   }
 
+  handlePickReply(m) {
+    this.setState({reply: null});
+    if (m) {
+      let cont = m.content
+      if (cont) {
+        if (typeof cont == 'string') {
+          cont = Drafty.init(cont);
+        }
+        cont = Drafty.replyPreview(cont, 30, quotePreviewFmt);
+
+        // Get the author.
+        let header;
+        const seq = m.seq;
+        const topic = this.props.tinode.getTopic(this.state.topic);
+        const m0 = topic.findMessage(seq);
+        let thisFrom;
+        if (m0) {
+          thisFrom = m0.from || 'chan';
+          if (thisFrom != this.props.myUserId) {
+            const topic = this.props.tinode.getTopic(this.state.topic);
+            const user = topic.userDesc(thisFrom);
+            if (user && user.public) {
+              header = user.public.fn;
+            }
+          } else {
+            header = this.props.myUserName;
+          }
+        }
+
+        header = Drafty.init(header || '');
+        // Make small image previews.
+        const ents = [];
+        if (cont.ent) {
+          Drafty.entities(cont, (data, idx, tp) => {
+            if (tp == 'IM') {
+              ents.push({
+                tp: tp,
+                data: data
+              });
+            }
+
+          }, this);
+        }
+        this.convertIntoThumbnails(ents, 0, (success) => {
```

Comment:
This looks un-javascripty to me. You can chain promises without recursion. 

Assuming `do_work()` scales the image and returns a promise, something like this would work for serial execution:
```js
for (let i = 0, p = Promise.resolve(); i < ents.length; i++) {
    p = p.then(() => do_work(ents[i]));
}
```
Or for parallel:

```js
const promises = [];
ents.map((ent) => {
    promises.push(do_work(ent));
});
Promise.all(promises).then(/* handle completion of all promises */);
```